packages/swapper/src/swappers/JupiterSwapper/endpoints.ts

```diff
@@ -0,0 +1,144 @@
+import { CHAIN_NAMESPACE, fromAssetId, solAssetId } from '@shapeshiftoss/caip'
+import type { BuildSendApiTxInput, GetFeeDataInput } from '@shapeshiftoss/chain-adapters'
+import type { SolanaSignTx } from '@shapeshiftoss/hdwallet-core'
+import type { KnownChainIds } from '@shapeshiftoss/types'
+import { PublicKey, type TransactionInstruction } from '@solana/web3.js'
+import type { AxiosError } from 'axios'
+
+import type { GetUnsignedSolanaTransactionArgs } from '../../types'
+import { type SwapperApi } from '../../types'
+import { checkSolanaSwapStatus, isExecutableTradeQuote, isExecutableTradeStep } from '../../utils'
+import type { Instruction } from './models/Instruction'
+import { getTradeQuote, getTradeRate } from './swapperApi/getTradeQuote'
+import { getJupiterSwapInstructions } from './utils/helpers'
+
+export const jupiterApi: SwapperApi = {
+  getTradeQuote,
+  getTradeRate,
+  getUnsignedSolanaTransaction: async ({
+    tradeQuote,
+    from,
+    assertGetSolanaChainAdapter,
+    config,
+  }: GetUnsignedSolanaTransactionArgs): Promise<SolanaSignTx> => {
+    if (!isExecutableTradeQuote(tradeQuote)) throw Error('Unable to execute trade')
+
+    const jupiterUrl = config.REACT_APP_JUPITER_API_URL
+
+    const step = tradeQuote.steps[0]
+
+    if (!isExecutableTradeStep(step)) throw Error('Unable to execute step')
+
+    if (!tradeQuote.rawQuote) throw Error('Missing raw quote')
+
+    const contractAddress =
+      step.buyAsset.assetId === solAssetId
+        ? undefined
+        : fromAssetId(step.buyAsset.assetId).assetReference
+
+    const adapter = assertGetSolanaChainAdapter(step.sellAsset.chainId)
+
+    const isManualReceiveAddress = tradeQuote.receiveAddress !== from
```

Comment:
Manual address is very specific, it has multiple edge cases that I had to handle:

- We are not using wallet address but token ATA, so we need to create these ATA if they don't exist, and use them in the jupiter swap instructions endpoints
- If it's Token2022 or SPL, this ATA derivation changes
- If we send SOL with manual receive address, we don't support it because Jupiter only supports sending wSOL under the hood with external address

Reply:
Renamed to `isCrossAccountTrade` in previous commits and plumbing cleaned up, closing this